File: samples/java_layout/src/main/java/io/customer/android/sample/java_layout/sdk/InAppMessageEventListener.java
package io.customer.android.sample.java_layout.sdk;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import io.customer.android.sample.java_layout.utils.Logger;
import io.customer.messaginginapp.type.InAppEventListener;
import io.customer.messaginginapp.type.InAppMessage;
import io.customer.sdk.CustomerIO;

/**
 * Sample implementation for {@link InAppEventListener}.
 */
public class InAppMessageEventListener implements InAppEventListener {
    private final Logger logger;

    public InAppMessageEventListener(Logger logger) {
        this.logger = logger;
    }

    @Override
    public void messageShown(@NonNull InAppMessage message) {
        logInAppEvent("in-app message: messageShown. message: %s", message);
        trackInAppEvent("messageShown", message);
    }

    @Override
    public void messageDismissed(@NonNull InAppMessage message) {
        logInAppEvent("in-app message: messageDismissed. message: %s", message);
        trackInAppEvent("messageDismissed", message);
    }

    @Override
    public void errorWithMessage(@NonNull InAppMessage message) {
        logInAppEvent("in-app message: errorWithMessage. message: %s", message);
        trackInAppEvent("errorWithMessage", message);
    }

    @Override
    public void messageActionTaken(@NonNull InAppMessage message, @NonNull String actionValue, @NonNull String actionName) {
        logInAppEvent("in-app message: messageActionTaken. action: %s, name: %s, message: %s", actionValue, actionName, message);
        trackInAppEvent("messageActionTaken",
                message,
                new HashMap<String, String>() {{
                    put("action-value", actionValue);
                    put("action-name", actionName);
                }});
    }

    private void trackInAppEvent(@NonNull String eventName, @NonNull InAppMessage message) {
        trackInAppEvent(eventName, message, null);
    }

    private void logInAppEvent(@NonNull String format, @NonNull Object... args) {
        logger.v(String.format(Locale.ENGLISH, format, args));
    }

    private void trackInAppEvent(@NonNull String eventName, @NonNull InAppMessage message, @Nullable Map<String, String> arguments) {
        CustomerIO.instance().track(
                "in-app message action",
                new HashMap<String, String>() {{
                    if (arguments != null) {
                        putAll(arguments);
                    }
                    put("event-name", eventName);
                    put("message-id", message.getMessageId());
                    String deliveryId = message.getDeliveryId();
                    put("delivery-id", deliveryId == null ? "NULL" : deliveryId);
                }}
        );
    }
}

File: samples/kotlin_compose/src/main/java/io/customer/android/sample/kotlin_compose/data/sdk/InAppMessageEventListener.kt
package io.customer.android.sample.kotlin_compose.data.sdk

import io.customer.android.sample.kotlin_compose.util.Logger
import io.customer.messaginginapp.type.InAppEventListener
import io.customer.messaginginapp.type.InAppMessage
import io.customer.messaginginapp.type.InAppMessageError
import io.customer.sdk.CustomerIO

/**
 * Sample implementation for `InAppEventListener`.
 */
class InAppMessageEventListener(private val logger: Logger = Logger()) : InAppEventListener {

    override fun messageShown(message: InAppMessage) {
        logInAppEvent("in-app message: messageShown. message: $message")
        trackInAppEvent("messageShown", message)
    }

    override fun messageDismissed(message: InAppMessage) {
        logInAppEvent("in-app message: messageDismissed. message: $message")
        trackInAppEvent("messageDismissed", message)
    }

    // Still a required member of the interface. The SDK always calls the overload below, so this
    // only runs if something reports a failure without classifying it.
    override fun errorWithMessage(message: InAppMessage) {
        logInAppEvent("in-app message: errorWithMessage (no reason). message: $message")
        trackInAppEvent("errorWithMessage", message)
    }

    override fun errorWithMessage(message: InAppMessage, error: InAppMessageError) {
        logInAppEvent("in-app message: errorWithMessage. reason: ${error.reason}, detail: ${error.detail}, message: $message")
        trackInAppEvent(
            "errorWithMessage",
            message,
            hashMapOf(
                "error-reason" to error.reason.name,
                "error-detail" to (error.detail ?: "NULL"),
                "error-code" to (error.code?.toString() ?: "NULL")
            )
        )
    }

    override fun messageActionTaken(
        message: InAppMessage,
        actionValue: String,
        actionName: String
    ) {
        logInAppEvent("in-app message: messageActionTaken. action: $actionValue, name: $actionName, message: $message")
        trackInAppEvent(
            "messageActionTaken",
            message,
            hashMapOf("action-value" to actionValue, "action-name" to actionName)
        )
    }

    private fun trackInAppEvent(eventName: String, message: InAppMessage) {
        trackInAppEvent(eventName, message, null)
    }

    private fun logInAppEvent(message: String) {
        logger.v(message)
    }

    private fun trackInAppEvent(
        eventName: String,
        message: InAppMessage,
        arguments: Map<String, String>?
    ) {
        CustomerIO.instance().track(
            "in-app message action",
            HashMap<String, String>().apply {
                arguments?.let { putAll(it) }
                put("event-name", eventName)
                put("message-id", message.messageId)
                put("delivery-id", message.deliveryId ?: "NULL")
            }
        )
    }
}
